flows/teamleader.py

```diff
@@ -0,0 +1,132 @@
+from time import sleep
+from typing import Any, Union, cast
+from datetime import datetime
+
+from prefect import task, get_run_logger
+import requests
+from requests import Response
+
+from models import *
+from database import *
+
+
+class TeamleaderRequestException(Exception):
+    pass
+
+
+@task
+def refresh_auth_token(conn: Connection, auth: TL_Auth) -> TL_Auth:
+    logger = get_run_logger()
+    logger.info("Refreshing access token")
+
+    # Check if the database is in a valid state
+    validate_db_auth_state(conn)
+
+    # Refresh the authorization token
+    response = requests.post(
+        auth.uri + "/access_token",
+        data={
+            "client_id": auth.client_id,
+            "client_secret": auth.client_secret.get_secret_value(),
+            "refresh_token": auth.refresh_token.get_secret_value(),
+            "grant_type": "refresh_token",
+        },
+    )
+    if response.status_code != 200:
+        raise Exception(
+            f"Could not refresh token. Status code {response.status_code} - {response.reason}"
+        )
+    response = response.json()
+
+    auth = TL_Auth(
+        uri=auth.uri,
+        client_id=auth.client_id,
+        client_secret=auth.client_secret,
+        refresh_token=response["refresh_token"],  # Does this need a SecretStr() ?
+        access_token=response["access_token"],  # idem
+    )
+
+    save_tokens_to_database(auth, conn)
+    logger.info("Updated access token and refresh token in database.")
+    return auth
+
+
+def get_request_headers(auth: TL_Auth):
+    return {"Authorization": f"Bearer {auth.access_token.get_secret_value()}"}
+
+
+def request_teamleader_info(
+    req: TL_RequestInfo,
+    auth: TL_Auth,
+    conn: Connection,
+) -> Tuple[TL_ResponseInfo, TL_Auth]:
+    response, auth = request_teamleader(req, auth, conn)
+    return (
+        TL_ResponseInfo(
+            resource=response.resource,
+            ratelimit_remaining=response.ratelimit_remaining,
+            ratelimit_reset=response.ratelimit_reset,
+            data=cast(dict[str, Any], response.data),
+        ),
+        auth,
+    )
+
+
+def request_teamleader_list(
```

Comment:
I think it would make more sense to make this into a task rather than the underlying post request, it makes issues in the flow much more visible. e.g. if a call fails you see that it is a request to teamleader failing instead of 'some' post request.